src/github.com/getlantern/netx/copy.go

```diff
@@ -32,29 +35,43 @@ func doCopy(dst net.Conn, src net.Conn, buf []byte, errCh chan error, stop *uint
 
 	for {
 		if atomic.LoadUint32(stop) == 1 {
-			break
+			return
 		}
 		deadline := time.Now().Add(copyTimeout)
 		src.SetReadDeadline(deadline)
 		dst.SetWriteDeadline(deadline)
 		nr, er := src.Read(buf)
 		if nr > 0 {
-			nw, ew := dst.Write(buf[0:nr])
-			if ew != nil && !isTimeout(err) {
+			ew := writeTo(dst, buf[0:nr], writeTimeout)
+			if ew != nil {
 				err = ew
-				break
-			}
-			if nr != nw {
-				err = io.ErrShortWrite
-				break
+				return
 			}
 		}
 		if er == io.EOF {
-			break
+			return
 		}
 		if er != nil && !isTimeout(er) {
 			err = er
-			break
+			return
+		}
+	}
+}
+
+func writeTo(dst net.Conn, buf []byte, writeTimeout time.Duration) error {
```

Comment:
@fffw I think this addresses your concern about short writes.  I modified the test to make sure we cover this.

Your idea of just setting the deadline after the other pipe has finished is interesting, but I'm not sure if it will work on all platforms.  Specifically, I don't know if calls to SetWriteDeadline() always affect existing Write() calls.  They do on Mac, but that seems like the sort of thing that might be platform specific.

Using an atomic closed indicator here seems to work.  Are you okay with that?